### PR TITLE
Port fill_ and zero_ to call Aten (CPU)

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Large tensor 3,000,000 elements:

Before

```
In [50]: timeit y = x.zero_()                                                                                                                                                                                                                                                              
659 µs ± 16.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [51]: timeit y = x.zero_()                                                                                                                                                                                                                                                              
637 µs ± 4.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [52]: timeit y = x.zero_()                                                                                                                                                                                                                                                              
624 µs ± 5.78 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

After

```
In [8]: timeit y = x.zero_()                                                                                                                                                                                                                                                               
608 µs ± 838 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [9]: timeit y = x.zero_()                                                                                                                                                                                                                                                               
605 µs ± 6.84 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [10]: timeit y = x.zero_()                                                                                                                                                                                                                                                              
611 µs ± 4.65 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Small tensors adds classical 300ns overhead of TensorIterator, we are working on reducing it.

### CODE REVIEW

**Comment by zou3519** [comment]
(some initial questions) How large is the benchmarked inputs? Could you also benchmark two things: (1) a very small input of size 1 (to see overhead), and (2) a larger input (to see any perf changes)

**Comment by VitalyFedyunin** [comment]
> (some initial questions) How large is the benchmarked inputs? Could you also benchmark two things: (1) a very small input of size 1 (to see overhead), and (2) a larger input where overhead doesn't dominate (to see any perf changes)

Description updated. 

Small tensors adds classical 300ns overhead. 

Before

```
In [60]: x = torch.randn(1)                                                                                                                                                                                                                                                                
In [61]: timeit y = x.zero_()                                                                                                                                                                                                                                                              
681 ns ± 11 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

After

```
In [1]: x = torch.randn(20)                                                                                                                                                                                                                                                                
In [2]: timeit y = x.zero_()                                                                                                                                                                                                                                                               
936 ns ± 18.6 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

File: aten/src/ATen/native/Fill.cpp [hunk 1]
```diff
@@ -27,6 +27,18 @@ Tensor& fill_(Tensor& self, const Tensor& value) {
 
 DEFINE_DISPATCH(fill_stub);
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  zero ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Tensor& zero_out(Tensor& self) {
+  auto iter = TensorIterator::nullary_op(self);
```
Comment (zou3519) [inline aten/src/ATen/native/Fill.cpp]: Could this just be `native::fill_out(self, 0)` to avoid code duplication?

File: aten/src/ATen/native/Fill.cpp [hunk 2]
```diff
@@ -27,6 +27,18 @@ Tensor& fill_(Tensor& self, const Tensor& value) {
 
 DEFINE_DISPATCH(fill_stub);
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  zero ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Tensor& zero_out(Tensor& self) {
```
Comment (zou3519) [inline aten/src/ATen/native/Fill.cpp]: Is `zero_out` used anywhere?

**Review by zou3519 — COMMENTED** [review]
(not full review yet)

File: aten/src/TH/generic/THTensorConv.cpp [hunk 3]
```diff
@@ -623,7 +623,6 @@ void THTensor_(conv2DRevger)(THTensor *r_, scalar_t beta, scalar_t alpha, THTens
 
   if (nelem == 0 || beta == 0 || nelem != THTensor_(nElement)(r_))
   {
-    /*THTensor_(zero)(r_);*/
```
Comment (zou3519) [inline aten/src/TH/generic/THTensorConv.cpp]: (just a comment) Why were these here before? I feel like there's some interesting history here

File: aten/src/TH/generic/THTensorConv.cpp [hunk 4]
```diff
@@ -1252,8 +1246,11 @@ void THTensor_(conv2Dmul)(THTensor *r_, scalar_t beta, scalar_t alpha, THTensor
 
   nelem = THTensor_(nElement)(r_);
   THTensor_(resize2d)(r_, nOutputRows, nOutputCols);
-  if (nelem == 0 || beta == 0 || nelem != THTensor_(nElement)(r_))
-    THTensor_(zero)(r_);
+  if (nelem == 0 || beta == 0 || nelem != THTensor_(nElement)(r_)) {
+    at::Tensor r_wrap = THTensor_wrap(r_);
+    r_wrap.zero_();
```
Comment (zou3519) [inline aten/src/TH/generic/THTensorConv.cpp]: This incurs an extra dispatch. Last time I measured, dispatch overhead was ~200ns. It *probably* doesn't matter too much

Comment (zou3519) [reply aten/src/TH/generic/THTensorConv.cpp]: Also, we're repeating
```
at::Tensor r_wrap = THTensor_wrap(r_);
r_wrap.zero_();
```
a lot. Maybe it's better to write a `THTensor_fill(TensorImpl*)` helper function?

As an alternative, this could just be a one-liner: `THTensor_wrap(r_).zero_();`.

**Review by zou3519 — COMMENTED** [review]
Had some minor comments. I'm not sure how much to worry about the additional overhead:
- there's the extra 300ns of TensorIterator overhead
- IIRC THTensor does no dispatching. Changing the calls to use Tensor::zero_ / Tensor::fill_ makes it so that we dispatch more and incur some dispatching overhead.

**Review by gchanan — CHANGES_REQUESTED** [review]
ya, I'd just use the fill implementation and either put the wrap/fill in one line or keep THTensor_fill and have it just do the one-line wrap/fill.
